Use temperature of current preset when set fritz HVAC mode to HEAT

naruxde opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
I use the FRITZ! DECT 302 thermostats on my heaters. In the Fritz! Box is a heating plan stored, which sets the temperature to COMFORT or ECO at certain times of the day. If I switch off the heating `HVACMode.OFF` via Home Assistant and turn it on again, the COMFORT temperature is always set, although the schedule for ECO is still active.

This change is intended to use the ECO temperature when switching on, if the schedule for ECO is still active, otherwise the COMFORT temperature.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://www.home-assistant.io/integrations/fritzbox

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

naruxde committed:
Use temperature of current preset when set fritz HVAC mode to HEAT

If the HVAC mode of the Fritzbox thermostats changes from `HVACMode.OFF`
to `HVAMode.HEAT`, the current preset (COMFORT / ECO) should be
observed. Depending on the status of the current preset, the set
temperature of comfort / eco is set as the new temperature.

mib1185 commented on tests/components/fritzbox/test_climate.py:
```diff
@@ -376,6 +376,28 @@ async def test_set_hvac_mode_heat(hass: HomeAssistant, fritz: Mock) -> None:
+        hass, MOCK_CONFIG[FB_DOMAIN][CONF_DEVICES][0], ENTITY_ID, device, fritz
+    )
+
+    device.nextchange_temperature = device.comfort_temperature
```
```suggestion
```
you already set the next change temp above

mib1185 reviewed (changes requested):
Hi @naruxde

thanks for this contribution 👍 
Besides the small comment from my side, the test fails - please check this. Don't hesitate to call for help if needed 😉 

regards,
Michael

naruxde commented:
Hi @mib1185!

I'm really sorry about what always went wrong 😖. I had even, before the push, made a pytest with `call(17)` on my system to see if it really fails.

mib1185 reviewed (changes requested):
we now have a regression with tests `test_set_temperature_mode_heat` and  `test_set_hvac_mode_heat` 🧐 
i've not checked the code yet in detail, so would be grate if you could have a look first.

naruxde commented:
Yes, I'm just looking at that. We have with the `FritzDeviceClimateMock` a `comfort_temperature = 22.0` and an `eco_temperature = 16.0`. Since `nextchange_temperature = 22.0`, `value_scheduled_preset` assumes that we are in `PRESET_ECO`.

So we would have to change `nextchange_temperature` in the `FritzDeviceClimateMock` class if we want to be in `PRESET_COMFORT` or adjust the value in the old `test_set_hvac_mode_heat` and remove it from the new `test_set_hvac_mode_heat_eco`.

We then generally only have the problem if eco and comfort are set to the same values in the Fritzbox, which mode should we use in Home Assistant. 🤔 - Aktuell wäre es `PRESET_ECO`, was ist nicht unbedingt falsch finde 😆

naruxde commented:
I pushed a commit as a suggestion for changing the tests. But maybe you have other ideas 🙂

mib1185 reviewed (changes requested):
I think using `value_scheduled_preset` is the wrong approach, since it intention was another. It will always return "eco mode", if `nextchange_temperature` is not the `eco_temperature`, even when `nextchange_temperature` is 128 (_or 255_) (_which means not set_)

mib1185 commented on homeassistant/components/fritzbox/climate.py:
```diff
@@ -174,6 +175,8 @@ async def async_set_hvac_mode(self, hvac_mode: HVACMode) -> None:
             return
         if hvac_mode == HVACMode.OFF:
             await self.async_set_temperature(temperature=OFF_REPORT_SET_TEMPERATURE)
+        elif value_scheduled_preset(self.data) is PRESET_ECO:
+            await self.async_set_temperature(temperature=self.data.eco_temperature)
         else:
             await self.async_set_temperature(temperature=self.data.comfort_temperature)
```
i think we have a logical issue in [`async_set_temperature`](https://github.com/home-assistant/core/blob/7ada2f864cb7fc68d2576d979af5c21d33859c37/homeassistant/components/fritzbox/climate.py#L136-L146) it ignores any given temperature, when a hvac mode is provided - this seems to be wrong

mib1185 replied:
will discuss with other core members and report back, as soon as i've an answer ...

naruxde committed:
fixup do not use value_scheduled_preset

Co-authored-by: Michael <35783820+mib1185@users.noreply.github.com>

mib1185 commented:
please avoid rebase + force-push. this makes hard to get what has changed since last review, thx 👍

mib1185 commented:
Hi @naruxde 
now as #126072 has been merged, you need to rebase your branch to latest `dev` (_yes, in this case its ok, as soon as it is only a rebase, without code changes_ 😉) and go ahead with your PR, thx 👍

naruxde committed:
Add current_preset value to test_set_hvac_mode

The current_preset parameter allows the mock to be set to an active
preset. When setting HVACMode.HEAT, the respective temperature of the
ECO/COMFORT preset should be set.

mib1185 reviewed (commented):
Hi @naruxde 

I thought about it again and unfortunately came to the conclusion, that all the "next change " sensors needs to be fixed. The api returns the comfort temperature as next temperature, when there is no scheduler configured. The only way to detect if a scheduler is configured, is to check the next change time - if it is 0, than there is no scheduler, else it contains a valid future timestamp. I'm already working on a further PR to fix this. Afterwards we should really be able to go ahead with your PR.

thx for your patients :+1:

naruxde commented:
Hi @mib1185!

Improvements are always good! I would also have no problem with it if you integrate the "hack" from this PR directly into your PR, which maybe results in a better solution. What ever you like most.

I look forward to it when the problem described above is solved, gladly also with a better solution than this 😆.

mib1185 committed:
Merge branch 'dev' into fritz-hvac-preset-mode

mib1185 commented on homeassistant/components/fritzbox/climate.py:
add the import `from .sensor import value_scheduled_preset`

mib1185 commented on homeassistant/components/fritzbox/climate.py:
```diff
@@ -177,7 +177,13 @@ async def async_set_hvac_mode(self, hvac_mode: HVACMode) -> None:
             await self.async_set_temperature(temperature=OFF_REPORT_SET_TEMPERATURE)
         else:
-            await self.async_set_temperature(temperature=self.data.comfort_temperature)
+            if (
+                nextchange_temperature := self.data.nextchange_temperature
+            ) is not None and nextchange_temperature == self.data.comfort_temperature:
```
```suggestion
            if value_scheduled_preset(self.data) == PRESET_ECO:
```

mib1185 commented on tests/components/fritzbox/test_climate.py:
```diff
@@ -313,24 +313,35 @@ async def test_set_temperature(
-        ({ATTR_HVAC_MODE: HVACMode.HEAT}, 0.0, [call(22)]),
-        ({ATTR_HVAC_MODE: HVACMode.HEAT}, 18, []),
-        ({ATTR_HVAC_MODE: HVACMode.HEAT}, 22, []),
+        ({ATTR_HVAC_MODE: HVACMode.OFF}, 22, PRESET_COMFORT, [call(0)]),
+        ({ATTR_HVAC_MODE: HVACMode.HEAT}, 0.0, PRESET_ECO, [call(16)]),
+        ({ATTR_HVAC_MODE: HVACMode.OFF}, 16, PRESET_ECO, [call(0)]),
+        ({ATTR_HVAC_MODE: HVACMode.HEAT}, 0.0, PRESET_COMFORT, [call(22)]),
+        ({ATTR_HVAC_MODE: HVACMode.HEAT}, 18, PRESET_ECO, []),
+        ({ATTR_HVAC_MODE: HVACMode.HEAT}, 22, PRESET_COMFORT, []),
```
```suggestion
        # mode off always sets target temperature to 0
        ({ATTR_HVAC_MODE: HVACMode.OFF}, 22, PRESET_COMFORT, [call(0)]),
        ({ATTR_HVAC_MODE: HVACMode.OFF}, 16, PRESET_ECO, [call(0)]),
        ({ATTR_HVAC_MODE: HVACMode.OFF}, 16, None, [call(0)]),
        # mode heat sets target temperature based on current scheduled preset,
        # when not already in mode heat
        ({ATTR_HVAC_MODE: HVACMode.HEAT}, 0.0, PRESET_COMFORT, [call(22)]),
        ({ATTR_HVAC_MODE: HVACMode.HEAT}, 0.0, PRESET_ECO, [call(16)]),
        ({ATTR_HVAC_MODE: HVACMode.HEAT}, 0.0, None, [call(22)]),
        # mode heat does not set target temperature, when already in mode heat
        ({ATTR_HVAC_MODE: HVACMode.HEAT}, 16, PRESET_COMFORT, []),
        ({ATTR_HVAC_MODE: HVACMode.HEAT}, 16, PRESET_ECO, []),
        ({ATTR_HVAC_MODE: HVACMode.HEAT}, 16, None, []),
        ({ATTR_HVAC_MODE: HVACMode.HEAT}, 22, PRESET_COMFORT, []),
        ({ATTR_HVAC_MODE: HVACMode.HEAT}, 22, PRESET_ECO, []),
        ({ATTR_HVAC_MODE: HVACMode.HEAT}, 22, None, []),
```

mib1185 commented on tests/components/fritzbox/test_climate.py:
```diff
@@ -313,24 +313,35 @@ async def test_set_temperature(
     device = FritzDeviceClimateMock()
     device.target_temperature = target_temperature
+
+    # Configure the mock to requested preset
+    if current_preset is PRESET_COMFORT:
+        device.nextchange_temperature = device.eco_temperature
+    else:
+        # This is the default value of FritzDeviceClimateMock
+        device.nextchange_temperature = device.comfort_temperature
```
```suggestion
    if current_preset is PRESET_COMFORT:
        device.nextchange_temperature = device.eco_temperature
    elif current_preset is PRESET_ECO:
        device.nextchange_temperature = device.comfort_temperature
    else:
        device.nextchange_endperiod = 0
```

mib1185 reviewed (changes requested):
Hi @naruxde 
we're close to get this ready and no, i won't put this into some of my PRs, because it is your contribution :wink: 
Please apply the following comments, thx :+1:

naruxde committed:
fixup Use the updated value_scheduled_preset function

To distinguish which temperature should be used when setting the
`HVAMode.HEAT`, `value_schedules_preset` is now used again, which has
been updated since the first commit. If no schedule is active, the
comfort_temperature is used. Otherwise, the respective temperature of
the current preset.

Co-authored-by: Michael <35783820+mib1185@users.noreply.github.com>

mib1185 reviewed (commented):
don't forget to click on "Ready to review" (_https://github.com/home-assistant/core/pull/126044#issuecomment-2352974042_) 😉

naruxde commented:
Had to test locally 😋

mib1185 reviewed (approved):
LGTM, thx @naruxde 👍
